pkg/extensions/tbtc/tbtc.go

```diff
@@ -36,29 +38,73 @@ func Initialize(ctx context.Context, handle Handle) error {
 		)
 	}
 
+	err = tbtc.monitorProvideRedemptionSignature(
+		ctx,
+		exponentialBackoff,
+		105*time.Minute, // 15 minutes before the 2 hours on-chain timeout
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"could not initialize provide redemption "+
+				"signature monitoring: [%v]",
+			err,
+		)
+	}
+
 	logger.Infof("tbtc extension has been initialized")
 
 	return nil
 }
 
-type depositEventHandler func(deposit string)
+type depositEvent interface {
+	getDepositAddress() string
+}
+
+type baseDepositEvent struct {
+	depositAddress string
+}
+
+func (bdd *baseDepositEvent) getDepositAddress() string {
+	return bdd.depositAddress
+}
+
+type depositRedemptionRequestedEvent struct {
+	*baseDepositEvent
+
+	digest      [32]uint8
+	blockNumber uint64
+}
+
+type depositEventHandler func(depositEvent)
 
 type watchDepositEventFn func(
 	handler depositEventHandler,
 ) (subscription.EventSubscription, error)
 
-type watchKeepClosedFn func(deposit string) (
+type watchKeepClosedFn func(depositAddress string) (
 	keepClosedChan chan struct{},
 	unsubscribe func(),
 	err error,
 )
 
-type submitDepositTxFn func(deposit string) error
+type depositMonitoringContext struct {
+	startEvent depositEvent
+}
+
+type submitDepositTxFn func(*depositMonitoringContext) error
 
 type backoffFn func(iteration int) time.Duration
```

Comment:
We have more and more abstractions added here and it's easy to lose track. Let's keep the definitions used in `monitorAndAct` right before `monitorAndAct`:
```
type watchDepositEventFn func(
	handler depositEventHandler,
) (subscription.EventSubscription, error)

type watchKeepClosedFn func(depositAddress string) (
	keepClosedChan chan struct{},
	unsubscribe func(),
	err error,
)

type submitDepositTxFn func(*depositMonitoringContext) error

type backoffFn func(iteration int) time.Duration

// TODO:
//  1. Filter incoming events by operator interest.
//  2. Incoming events deduplication.
//  3. Resume monitoring after client restart.
func (t *tbtc) monitorAndAct(
	ctx context.Context,
	monitoringName string,
	monitoringStartFn watchDepositEventFn,
	monitoringStopFn watchDepositEventFn,
	keepClosedFn watchKeepClosedFn,
	actFn submitDepositTxFn,
	actBackoffFn backoffFn,
	timeout time.Duration,
) (subscription.EventSubscription, error) {
```

Reply:
Even with the current version of the code where we got rid of the additional abstractions, I still opt for moving those functions to live next to `monitorAndAct`.